File: backend/src/app/geoip/crowdsec.py
"""CrowdSec bad-actor blocklist (DR-G8) — stream-mode bouncer, process cache.

A scheduler job pulls ban decisions from the CrowdSec LAPI
(``/v1/decisions/stream``): the first call after process start fetches the
full set (``startup=true``), later calls only deltas. The middleware checks
IPs against the cached set — never live HTTP in the request path. A LAPI
outage keeps the last known bans (stale beats empty: dropping the list on a
hiccup would un-ban every attacker at once); the staleness is visible in the
status endpoint.

Single IPs (the overwhelming majority of community-blocklist entries) live in
an O(1) set; genuine ranges in a small tuple that is scanned linearly.
"""

from __future__ import annotations

import ipaddress
from datetime import UTC, datetime

import httpx
import structlog

from app.config import get_settings

log = structlog.get_logger("app.geoip")

_SYNC_TIMEOUT = 15.0

# Process-local state (single worker by design).
_banned_ips: set[str] = set()
_banned_ranges: dict[str, ipaddress.IPv4Network | ipaddress.IPv6Network] = {}
_startup_done = False
_last: dict = {"at": None, "ok": None, "detail": "never synced"}


def active() -> bool:
    """True when the blocklist check should run at all."""
    settings = get_settings()
    return settings.crowdsec_enabled and bool(settings.crowdsec_api_key)


def banned_count() -> int:
    return len(_banned_ips) + len(_banned_ranges)


def is_banned(ip: str) -> bool:
    """O(1) for single-IP bans, linear over the (few) range bans."""
    if ip in _banned_ips:
        return True
    if not _banned_ranges:
        return False
    try:
        addr = ipaddress.ip_address(ip)
    except ValueError:
        return False
    return any(addr in net for net in _banned_ranges.values() if addr.version == net.version)


def status() -> dict:
    return {
        "enabled": get_settings().crowdsec_enabled,
        "configured": active(),
        "banned_count": banned_count(),
        **_last,
    }


def _normalize(value: str) -> tuple[str, str] | None:
    """Decision value → ('ip', canonical) or ('range', canonical); None = junk."""
    try:
        net = ipaddress.ip_network(value.strip(), strict=False)
    except ValueError:
        return None
    if net.num_addresses == 1:
        return "ip", str(net.network_address)
    return "range", str(net)


def apply_decisions(new: list[dict], deleted: list[dict]) -> None:
    """Fold one stream delta into the cache (pure state transition, testable)."""
    for decision, removing in ((deleted, True), (new, False)):
        for d in decision:
            if (d.get("type") or "ban") != "ban":
                continue
            norm = _normalize(str(d.get("value", "")))
            if norm is None:
                continue
            kind, canonical = norm
            if removing:
                if kind == "ip":
                    _banned_ips.discard(canonical)
                else:
                    _banned_ranges.pop(canonical, None)
            elif kind == "ip":
                _banned_ips.add(canonical)
            else:
                _banned_ranges[canonical] = ipaddress.ip_network(canonical)


async def sync() -> None:
    """Scheduler entry point: pull the next stream delta (full set on first run)."""
    global _startup_done
    if not active():
        return
    settings = get_settings()
    url = settings.crowdsec_lapi_url.rstrip("/") + "/v1/decisions/stream"
    try:
        async with httpx.AsyncClient(timeout=_SYNC_TIMEOUT) as client:
            resp = await client.get(
                url,
                params={"startup": "true" if not _startup_done else "false"},
                headers={"X-Api-Key": settings.crowdsec_api_key},
            )
        if resp.status_code != 200:
            _finish(ok=False, detail=f"LAPI HTTP {resp.status_code}")
            return
        data = resp.json() or {}
    except (httpx.HTTPError, ValueError) as exc:
        # Keep the last known bans — stale beats empty (see module docstring).
        _finish(ok=False, detail=f"{exc.__class__.__name__}: {exc}")
        return
    apply_decisions(data.get("new") or [], data.get("deleted") or [])
    _startup_done = True
    _finish(ok=True, detail=f"{banned_count()} active bans")


def _finish(ok: bool, detail: str) -> None:
    _last.update(at=datetime.now(UTC).isoformat(), ok=ok, detail=detail)
    if ok:
        log.debug("geoip.crowdsec_sync", detail=detail)
    else:
        log.warning("geoip.crowdsec_sync_failed", detail=detail)

File: orbit/lib/orbit/geoip/crowdsec.ex
defmodule Orbit.GeoIP.Crowdsec do
  @moduledoc """
  CrowdSec bad-actor blocklist (DR-G8) — stream-mode bouncer, port of
  geoip/crowdsec.py. A 30s GenServer timer pulls ban decisions from the
  LAPI (`/v1/decisions/stream`): the first successful call fetches the full
  set (startup=true), later calls only deltas. The gate checks IPs against
  an ETS cache — never live HTTP in the request path. A LAPI outage keeps
  the last known bans (stale beats empty: dropping the list on a hiccup
  would un-ban every attacker at once); staleness is visible via `status/0`.

  Single IPs (the overwhelming majority of community-blocklist entries) are
  O(1) ETS members; genuine ranges (v4/v6) are scanned linearly.

  Same switch shape as GeoIP: configuring DASH_CROWDSEC_API_KEY turns it
  on, DASH_CROWDSEC_DISABLE=true turns it off without losing the key — and
  the blocklist bites even when the country restriction is off.
  """

  use GenServer

  require Logger

  alias Orbit.GeoIP.Rules

  @table :crowdsec_bans
  @sync_ms 30_000
  @sync_timeout_ms 15_000

  def start_link(opts) do
    GenServer.start_link(
      __MODULE__,
      Keyword.get(opts, :sync_on_start, true),
      name: Keyword.get(opts, :name, __MODULE__)
    )
  end

  @doc "True when the blocklist check should run at all."
  def active? do
    key = Application.get_env(:orbit, :crowdsec_api_key)
    key not in [nil, ""] and not Application.get_env(:orbit, :crowdsec_disable, false)
  end

  @doc "O(1) for single-IP bans, linear over the (few) range bans."
  def is_banned(ip) do
    :ets.member(@table, {:ip, ip}) or banned_by_range?(ip)
  rescue
    ArgumentError -> false
  end

  def banned_count do
    :ets.info(@table, :size)
  rescue
    ArgumentError -> 0
  end

  @doc "Sync health for status surfaces (python crowdsec.status parity)."
  def status(server \\ __MODULE__) do
    last =
      try do
        GenServer.call(server, :last)
      catch
        :exit, _ -> %{at: nil, ok: nil, detail: "not running"}
      end

    Map.merge(
      %{
        disabled: Application.get_env(:orbit, :crowdsec_disable, false),
        key_set: Application.get_env(:orbit, :crowdsec_api_key) not in [nil, ""],
        configured: active?(),
        banned_count: banned_count()
      },
      last
    )
  end

  @doc """
  Fold one stream delta into the ETS cache (pure state transition over the
  given table — testable). Only type=ban decisions count; junk values drop.
  """
  def apply_decisions(table \\ @table, new, deleted) do
    for {decisions, removing} <- [{deleted, true}, {new, false}],
        d <- decisions,
        (d["type"] || "ban") == "ban",
        norm = normalize(to_string(d["value"] || "")),
        norm != nil do
      case {norm, removing} do
        {{:ip, canonical}, true} -> :ets.delete(table, {:ip, canonical})
        {{:ip, canonical}, false} -> :ets.insert(table, {{:ip, canonical}})
        {{:range, canonical, _cidr}, true} -> :ets.delete(table, {:range, canonical})
        {{:range, canonical, cidr}, false} -> :ets.insert(table, {{:range, canonical}, cidr})
      end
    end

    :ok
  end

  # -- GenServer ------------------------------------------------------------

  @impl true
  def init(sync_on_start) do
    :ets.new(@table, [:named_table, :public, :set, read_concurrency: true])
    # First sync promptly (async — boot must not block on the LAPI).
    # sync_on_start=false is the test seam: Req.Test ownership must be
    # granted before the first pull.
    if sync_on_start, do: send(self(), :sync)
    {:ok, %{startup_done: false, last: %{at: nil, ok: nil, detail: "never synced"}}}
  end

  @impl true
  def handle_call(:last, _from, state), do: {:reply, state.last, state}

  @impl true
  def handle_info(:sync, state) do
    Process.send_after(self(), :sync, @sync_ms)

    if active?() do
      {:noreply, sync(state)}
    else
      {:noreply, state}
    end
  end

  defp sync(state) do
    url =
      Application.get_env(:orbit, :crowdsec_lapi_url, "http://crowdsec:8080")
      |> String.trim_trailing("/")
      |> Kernel.<>("/v1/decisions/stream")

    base_opts = [
      url: url,
      params: [startup: to_string(not state.startup_done)],
      headers: [{"x-api-key", Application.get_env(:orbit, :crowdsec_api_key)}],
      receive_timeout: @sync_timeout_ms,
      retry: false
    ]

    req_opts =
      case Application.get_env(:orbit, :crowdsec_req_plug) do
        nil -> base_opts
        plug -> Keyword.put(base_opts, :plug, plug)
      end

    case Req.get(req_opts) do
      {:ok, %{status: 200, body: body}} when is_map(body) ->
        apply_decisions(body["new"] || [], body["deleted"] || [])
        %{state | startup_done: true} |> finish(true, "#{banned_count()} active bans")

      {:ok, %{status: status}} ->
        # Keep the last known bans — stale beats empty (moduledoc).
        finish(state, false, "LAPI HTTP #{status}")

      {:error, error} ->
        finish(state, false, Exception.message(error))
    end
  end

  defp finish(state, ok, detail) do
    if ok do
      Logger.debug("geoip.crowdsec_sync detail=#{detail}")
    else
      Logger.warning("geoip.crowdsec_sync_failed detail=#{detail}")
    end

    %{state | last: %{at: DateTime.utc_now(), ok: ok, detail: detail}}
  end

  # Decision value → {:ip, canonical} | {:range, canonical, cidr} | nil.
  defp normalize(value) do
    case Rules.classify_entry(value) do
      {:cidr, {addr, prefix}} ->
        if prefix == max_prefix(addr) do
          {:ip, addr |> :inet.ntoa() |> to_string()}
        else
          canonical = "#{addr |> :inet.ntoa() |> to_string()}/#{prefix}"
          {:range, canonical, {addr, prefix}}
        end

      _ ->
        nil
    end
  end

  defp max_prefix(addr) when tuple_size(addr) == 4, do: 32
  defp max_prefix(addr) when tuple_size(addr) == 8, do: 128

  defp banned_by_range?(ip) do
    case :ets.match(@table, {{:range, :"$1"}, :"$2"}) do
      [] ->
        false

      ranges ->
        case :inet.parse_strict_address(String.to_charlist(ip)) do
          {:ok, addr} ->
            Enum.any?(ranges, fn [_canonical, cidr] -> Rules.addr_in_cidr?(addr, cidr) end)

          {:error, _} ->
            false
        end
    end
  end
end
